specs/impact-model-plugin.md

```diff
@@ -0,0 +1,335 @@
+
+The goal of the Impact Measurement Plugin (IMP) Spec is to define a common standardised interface to every IMM, current and future. 
+
+## Impact Model Interface Specification 
+
+The Impact Model Interface Specification project is here to standardise an interface to existing and future impact models (IMOs). The ecosystem of solutions we need for Impact measurement will sit on a solid bedrock of Impact Model Interfaces (IMIs) that have the same interface, to enable reusability, and comparability. 
+
+Any class library that exposes the IMI interface adheres to this specification.
+
+The interface describes the bidirectional communication between an IMO and what's calling the IMO. It describes the function signature, the data types and the exceptions used to pass back information to the caller.
+
+This version of the specification does not pick any specific language, but it does assume the language has support for exception handling.
+
+## Data Types
+
+### StaticParams
+
+This is a simple dictionary used to set up an IMM. Each IMM is different and can have a wide variety of parameters. The specification cannot define what these parameters will be, so we choose to have a simple data type of a dictionary which accepts all types and quantity of params.
+
+### Telemetry
+
+This is a unit of data to describe some inputs to an IMM. Since every IMM is different we can’t prescribe too much in the specification, however the only two fields that would be mandatory for each unit of telemetry is the date/time when the telemetry was gathered and the time-range for which the telemetry is for. For example, you might have some telemetry for CPU utilisation, but we also need to know when this telemetry was gathered and for what period of time does the telemetry measure.
+
+Telemetry can be aggregate or granular, for example it might contain an average of cpu utilization across a range of machines.
+
+`{ “date-time”: xxxx, “duration”: xxx, “cpu-util”: 0.5 }`
+
+Or it might contain more granular data for every machine like so:
+
+`{ “date-time”: xxxx, “duration”: xxx, machines: { “gsf-001”: “cpu-util”: 0.5, “gsf-002”:  “cpu-util”: 0.62 } }`
+
+Or it might contain even more granular data to the process level:
+
+`{ “date-time”: xxxx, “duration”: xxx, machines: { “gsf-001”: { processes: { “process-001”: “cpu-util”: 0.5 } }  }`
+
+This is just an example to demonstrate just how wide a variety of data there can be for Telemetry, there may be an infinite variety of inputs here. The Telemetry is really defined by the IMM, if the IMM does per processes modelling then the Telemetry can provide per process data.
+
+### TelemetryStream
+
+This data type returns a stream of Telemetry objects, used when we want to calculate impact metrics in real-time.
+
+### ImpactModel
+
+This datatype holds the instance for an IMO.
+
+### ImpactMetric

```

Comment:
We need to provide customer perspective of what an Impact metric is . For the end user (here in this case the software developer who is trying to measure the "impact" of his software), there should be a way for him/her to explicitly call out the impact metric they are trying to measure . Impact metric could be carbon, water, energy etc . We need to find a way to break this impact metric into some classes that can be leveraged directly by the customer on installing the Impact Measurement Plugin . This way, a person who is measuring carbon will be using classes that talk the "carbon" language (names of classes, methods etc) and someone who is measuring "water" will be using that appropriate modules.

As soon as the IMP is installed there should be a way to specify if the user is intending to measuring carbon, water, energy etc.  This should be given as input to the plugin along with the telemetry. 

For example 
pip install CloudCarbonFootprintplugin carbon

This will use the carbon related impact metric classes. The carbon related impact metric classes will probably derive from ImpactMetric class or interface.

Reply:
I suspect where we will end up is the impact will be determined by the plugin. So you'd install the plugin (IMP) for the specific method of measurement you want.

so com.thoughworks.ccf.carbon.sci
so com.thoughworks.ccf.carbon.ghg
so com.thoughworks.ccf.water

I suspect the whole methodology of the plugin, IMP, IMM will be fundamentally different for every approach. 

Although I don't think there is anything we are saying that would stop people from just writing a generic IMP that does many different things. I personally think they should be narrowly focused on a specific area.

If a plugin does do multiple impacts it could be configured with the params that are passed in when we instantiate this plugin (e.g. mode: ghg). But I do think we should somehow have some guidance where each plugin does one unique impact metric. Plugins that can be configured to return multiple impacts are going to be a huge headache down the line when people forgot to configure with the right impact mode and complain when one node returns water and another node returns sci.

Or actually maybe you need to declare at the top of your Impact Graph the type of Impact Metric you are measuring and then we can verify/validate that the plugins you are trying to us actually return that impact metric.